BUG: fix datetime.timedelta->timedelta64 unit detection logic

shoyer opened the pull request:
Fixes #11096

This existing logic (dating back seven years to Mark Weibe's original commit) seems to have an off-by-one error, and also omitted the base case -- which I think left `us_meta.base` undefined.

~~TODO: add a regression test for this, if possible.~~ Done.

eric-wieser commented on numpy/core/tests/test_datetime.py:
```diff
@@ -325,6 +325,15 @@ def test_timedelta_scalar_construction(self):
+                  datetime.timedelta(31)]
+        expected = np.array([28, 30, 31], dtype='timedelta64[D]')
+        actual = np.array(inputs, dtype='timedelta64[D]')
+        assert_equal(expected, actual)
```
This test doesn't fail for me on master

eric-wieser replied:
Ah, I remember now that this was transient

eric-wieser commented on numpy/core/tests/test_datetime.py:
```diff
@@ -325,6 +325,15 @@ def test_timedelta_scalar_construction(self):
         assert_raises(TypeError, np.timedelta64, a, 'm')
 
+    def test_timedelta_object_array_conversion(self):
+        # Regression test for GH11096
```
Nit: elsewhere we spell it `gh-11096`, with a hyphen

eric-wieser replied:
Looks like you caught another one of these in this file, not this one

eric-wieser commented:
This fix is unrelated to the issue. The only code path that comes close to calling this is `np.arange(<some timedelta args>)` - and I'm having trouble making that take the code path

shoyer committed:
TST: regression test for gh-6452

shoyer committed:
BUG: fix datetime.timedelta->timedelta64 unit detection logic

eric-wieser commented:
Looks like you just pushed the test by itself - I guess that was deliberate to see if it fails?

shoyer commented:
@eric-wieser yes, indeed....

eric-wieser commented:
Ah, you're right - my grepping missed the `*.c.src` files

shoyer commented:
And indeed, my regression test [fails on Travis](https://travis-ci.org/numpy/numpy/jobs/383039061):
```
_____________ TestDateTime.test_timedelta_object_array_conversion ______________
self = <numpy.core.tests.test_datetime.TestDateTime object at 0x7f140fa8e1d0>
    def test_timedelta_object_array_conversion(self):
        # Regression test for GH11096
        inputs = [datetime.timedelta(28),
                  datetime.timedelta(30),
                  datetime.timedelta(31)]
        expected = np.array([28, 30, 31], dtype='timedelta64[D]')
>       actual = np.array(inputs, dtype='timedelta64[D]')
E       TypeError: Cannot cast datetime.timedelta object from metadata [Y] to [D] according to the rule 'same_kind'
```

eric-wieser commented:
To make the test fail reliably on my machine, I have to do it with a single timedelta:

```
np.array(datetime.timedelta(weeks=1))
```

The bug is specific to when there are a whole number of weeks.

eric-wieser commented:
Looking more closely, it seems this code path's purpose is just to prevent the following two things:

```
np.array(datetime.timedelta(...), dtype='m8[Y]')
np.array(datetime.timedelta(...), dtype='m8[M]')
``` 
Right now, these fail for timedeltas that span whole weeks

shoyer commented:
> Can I request you add a test for one of these?

Yes, will  do.

eric-wieser commented:
I think this also fixes trying to convert a 0 timedelta too, `datetime.timedelta()`

shoyer committed:
TST: test for timedelta coersion

eric-wieser committed:
TST: Add extra regression tests for gh-11096

Conversion from timedelta was only reliably broken for integer numbers of weeks (including 0)

eric-wieser commented:
I've gone ahead and added two more tests - in particular, conversion was only ever broken for multiples of 7 days, which is why you need 28 in your list.

eric-wieser commented:
(restarting appveyor build)

eric-wieser commented:
Nice job tracking this down!
